Papers Please! Now with Markup!

WarlockD opened the pull request:
Closes #49902

## About The Pull Request

Since I heard about the pull and the reason for it I thought I would take a jab at fixing paper.  I am using the engine markedjs to parse but had to convert a bunch of the code to work with inferno, jsx and use our eslint.  I plan (MUCH later) to modify it to support old form's like we had and a simple macro to let you sign it.

As soon as the new ui_asset update happens, I will get the stamps to show up as background images to kind of mirror what it looks like in game

That being said, I am still new to react and web programing in general so putting this up early to get feedback and help with the interface.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

How can you NOT have paper in a game!  How can the hos take you seriously without a validated form?
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
fix: Added markdown for paper
<!-- 
add: Added new things
add: Added more things
del: Removed old things
tweak: tweaked a few things
balance: rebalanced something
fix: Added markdown for paper
soundadd: added a new sound thingy
sounddel: removed an old sound thingy
imageadd: added some icons and images
imagedel: deleted some icons and images
spellcheck: fixed a few typos
code: changed some code
refactor: refactored some code
config: changed some config setting
admin: messed with admin stuff
server: something server ops should know
-->
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

WarlockD committed:
Initial Paper changes

Lots of things done.  Now that I have the general idea on how this will work, going to clean it up a bit in js and fix some of the bugs

WarlockD committed:
Ok, so now we can make and save paper!

And a few other things.  Needs some twiking to the interface, I really think I do need a custom layout.

optimumtact commented:
Now this is some good shit

TiviPlus commented:
Based and speedmergepilled

Cyberboss commented:
Good PR

MrDoomBringer commented:
![image](https://user-images.githubusercontent.com/29008542/82099815-44396000-96d6-11ea-98a6-cfb46a7e119d.png)

ZeWaka commented:
re: #50969, can you either make clear that you're relicensing your changes to markedjs under the AGPL or keeping it under the original MIT license of the markedjs project?

Also a note: currently, the file you added, `tgui/packages/tgui/components/TextArea.js` is under an AGPL license, while the `tgui/packages/tgui/styles/components/TextArea.scss` file is under a MIT license, which is quite strange (and you're signing away the copyright to someone who is not you)

stylemistake commented on tgui/packages/tgui/interfaces/PaperSheet.js:
```diff
@@ -0,0 +1,71 @@
+    pen_color="black",
+    ...rest
+  } = data;
+  const [marked_value, setMarked] = useLocalState(context, text);
```
Not sure if this usage of `text` as a key for local state is intended.

```js
useLocalState(context, key, initialValue)
```

stylemistake commented on tgui/packages/tgui/styles/components/TextArea.scss:
```diff
@@ -0,0 +1,84 @@
+}
+
+/* Cut and paste to auto resize the textarea */
+.textarea-container > div, .textarea-container > textarea {
```
Replace these complex selectors with BEM-style simple selectors, which is:

```
.TextArea__baseline,
.TextArea__textarea {
  ...
}
```

If you need to reference that another class-less `<div ref={...} />`, just add a className to it, and add it here.

stylemistake commented on tgui/packages/tgui/styles/components/TextArea.scss:
```diff
@@ -0,0 +1,84 @@
+  width: 100%;
+}
+
+.textarea-container > textarea {
+  overflow: hidden;
+  position: absolute;
+  height: 100%;
+}
+
+.textarea-container > div {
+  padding-bottom: 1.5em; /* A bit more than one additional line of text. */
+  visibility: hidden;
+}
```
You can merge these blocks with the other ones below, after you fix the selectors.

stylemistake commented on tgui/packages/tgui/styles/themes/paper.scss:
```diff
@@ -0,0 +1,59 @@
+  }
+  .textarea-container {
+    position: relative;
+}
+.code {
+  font-family: Consolas,"courier new";
+  color: black;
+  background-color: #f1f1f1;
+  padding: 2px;
+  font-size: 85%;
+}
```
Bad indentation.

WarlockD commented:
Ah my bad, I basically cut and pasted the Input component and modified it
I will fix that.  Generaly anything I will do here will be MIT but I will put the TextAreaa under AGPL since its just a copy

stylemistake commented on tgui/packages/tgui/interfaces/PaperSheet.js:
```diff
@@ -0,0 +1,71 @@
+    setMarked(value);
+  };
+
+  const readSheet = marked_text => {
+    return (
+      <Box p="0px" textColor={pen_color} backgroundColor={pen_color}
+        width="auto" height={(window.innerHeight) + "px"}>
+        {marked(isFalsy(marked_text) ? '' : marked_text,
+          { breaks: true, smartypants: true })}
+      </Box>
+    );
+  };
+  const editSheet = () => {
+    return (
+      <Flex direction="column" justify="center">
+        <Flex.Item>
+          <Button.Confirm
+            content="FINISH"
+            onClick={() => act('save', { 'text': marked_value })}
+          />
+        </Flex.Item>
+        <Flex.Item>
+          <Flex>
+            <Flex.Item width="50%" >
+              <TextArea fluid height={(window.innerHeight-60) + "px"}
+                onInput={handleOnInput} />
+            </Flex.Item>
+            <Flex.Item width="50%">
+              {readSheet(marked_value)}
+            </Flex.Item>
+          </Flex>
+        </Flex.Item>
+      </Flex>
+    );
+  };
```
Please extract these functions into separate components. It's very easy in tgui 3.0, and it will separate concerns much better.

WarlockD replied:
Yea, originally that's how it was set up but I was having problems getting state to work across components.

stylemistake commented on tgui/packages/tgui/components/TextArea.js:
```diff
@@ -0,0 +1,184 @@
+        <div className="TextArea__baseline">
+          .
+        </div>
+        <div ref={this.fillerRef} />
```
Have you tried not using the filler div, but instead stretch textarea using:

```
position: absolute;
top: 0;
left: 0;
bottom: 0;
right: 0;
```

and this on `TextArea` class:

```
position: relative;
```

stylemistake reviewed (changes requested):
- [x] Move markedjs to `packages/marked`.
- [x] Put proper license headers to all marked files, maintain copyright notice from the marked project.
- [x] Put copyright on your own files and use your own name (can be Github username).
- [x] TextArea-related files must be MIT-licensed for the sake of consistency (it's part of base tgui), everything else license as you wish.

WarlockD commented:
Ugh I will put them all under MIT its just easyer for me

WarlockD committed:
Fixed copyright and moved folders around

Iamgoofball commented:
can we get some sweet pics of how this looks?

LemonInTheDark commented:
Absolute madman, I second the pic request.

Nebulacrity commented:
Damn, what a coding warlord.

spessman-007 commented:
@WarlockD I just cloned your repo and tried to run it. This is the first thing I see when I open a blank paper. I suggest using a white background.
![image](https://user-images.githubusercontent.com/62214048/82121149-8c6a8800-9793-11ea-9cd0-bfd0646d2acf.png)
For others, this is what happens when I click with a pen on the paper
![image](https://user-images.githubusercontent.com/62214048/82121165-a1471b80-9793-11ea-9538-1576d4159842.png)
Then I get asked to confirm.
![image](https://user-images.githubusercontent.com/62214048/82121175-b02dce00-9793-11ea-96d4-66fef5c30eda.png)
After that, well, the text is there.
![image](https://user-images.githubusercontent.com/62214048/82121178-b8860900-9793-11ea-842d-ee6c497b601c.png)

WarlockD commented:
Ugh yea trying to get the background to work is a bear.  I just woke up so fixing said issues.

WarlockD committed:
Updated Copyright and other fixes.

Moved stuff over to more components and switched to tabs.
Gonig to luch and be back to fix more stuff bleh.

WarlockD committed:
Refactoring text, bug fixes

Going to be trying the original string based markedjs because my rework is buggy.  Going to inject it into innerHTML and just pray

WarlockD committed:
Swaped my hacked version of markedjs with yarn

So yea, I gave up, to many escape bugs to fix.  Plus side it works fine!
Except for one small thing, we still have the same issue that the old paper has doh!  You can crash the server if you send the the book of alice in wonderland ( about 152k text file)

I can do some text file size limiting on the js, but if somone does an injection attack...

WarlockD commented:
![D2gdZpe](https://user-images.githubusercontent.com/1102671/82134844-d23e4500-97c1-11ea-902b-a506138353fb.gif)

WarlockD commented:
Right now we still have one problem.  The original problem honestly.  You can take the text of Alice in Wonderland (about 152k) and it will cause a weird pause, evey tick, on the server.  Time dilation goes all the way up to 140% on a locally run instance.  We hit a stack_trace crash:
````
[2020-05-17 01:19:29.290] runtime error: Maximum recursion level reached (perhaps there is an infinite loop)
 - To avoid this safety check, set world.loop_checks=0.
 - proc name: restrained (/mob/living/carbon/human/restrained)
 -   usr: (src)
 -   src: Elle Leach (/mob/living/carbon/human)
 -   src.loc: the floor (120,126,2) (/turf/open/floor/wood)
 -   call stack:
 - Elle Leach (/mob/living/carbon/human): restrained(0)
 - Elle Leach (/mob/living/carbon/human): incapacitated(0, 0, 0)
 - Elle Leach (/mob/living/carbon/human): shared ui interaction(the paper (/obj/item/paper))
 - Elle Leach (/mob/living/carbon/human): shared ui interaction(the paper (/obj/item/paper))
 - Elle Leach (/mob/living/carbon/human): default can use topic(the paper (/obj/item/paper))
 - /datum/ui_state/default (/datum/ui_state/default): can use topic(the paper (/obj/item/paper), Elle Leach (/mob/living/carbon/human))
 - the paper (/obj/item/paper): ui status(Elle Leach (/mob/living/carbon/human), /datum/ui_state/default (/datum/ui_state/default))
 - the paper (/obj/item/paper): ui status(Elle Leach (/mob/living/carbon/human), /datum/ui_state/default (/datum/ui_state/default))
 - /datum/tgui (/datum/tgui): update status(0)
 - /datum/tgui (/datum/tgui): push data(/list (/list), null, 0)
 - ...
 - the paper (/obj/item/paper): ui close(Elle Leach (/mob/living/carbon/human))
 - the paper (/obj/item/paper): ui close(Elle Leach (/mob/living/carbon/human))
 - the paper (/obj/item/paper): ui close(Elle Leach (/mob/living/carbon/human))
 - the paper (/obj/item/paper): ui close(Elle Leach (/mob/living/carbon/human))
 - the paper (/obj/item/paper): ui close(Elle Leach (/mob/living/carbon/human))
 - the paper (/obj/item/paper): ui close(Elle Leach (/mob/living/carbon/human))
 - the paper (/obj/item/paper): ui close(Elle Leach (/mob/living/carbon/human))
 - the paper (/obj/item/paper): ui close(Elle Leach (/mob/living/carbon/human))
 - Erodros (/client): uiclose("\[0x21010ace]")
````
So how can I solve this?  I can put an upper limit of, I don't know, 5000 characters, sanitize it  and it will be fine.  Is that good enough?

MrDoomBringer commented:
5000? Do like half that (If you're going to add a hard cap). Storing more information should be done through the use of multiple sheets of paper + clipboards or folders or something

stylemistake commented:
Standard A4 sheet of paper fits around 500 words, which is ~3000 characters.

WarlockD committed:
Limited to 3k, still have beyond issues.

Going to run the profiler tommow, I am passing out

WarlockD committed:
Added dompurify for a html sanatizer and a few fixes

WarlockD commented:
I think we are ready, I got it limited to 1000 characters and added a html sanitizer.

stylemistake commented:
Closes #49902

WarlockD committed:
Added copyright stuff

lordpidey commented on code/modules/paperwork/paper.dm:
```diff
@@ -108,201 +136,56 @@
-		if(user.is_literate())
-			user << browse("<HTML><HEAD><TITLE>[name]</TITLE></HEAD><BODY>[info_links]<HR>[stamps]</BODY><div align='right'style='position:fixed;bottom:0;font-style:bold;'><A href='?src=[REF(src)];help=1'>\[?\]</A></div></HTML>", "window=[name]")
+		if(finalized)
+			to_chat(user, "<span class='warning'>This sheet of paper has allready been written too!</span>")
```
Are these misspellings intentional?

WarlockD replied:
Sorry my dyslexia kicked in there, getting it fixed now

lordpidey commented on code/modules/paperwork/paper.dm:
```diff
@@ -108,201 +136,56 @@
 	if(istype(P, /obj/item/pen) || istype(P, /obj/item/toy/crayon))
-		if(user.is_literate())
-			user << browse("<HTML><HEAD><TITLE>[name]</TITLE></HEAD><BODY>[info_links]<HR>[stamps]</BODY><div align='right'style='position:fixed;bottom:0;font-style:bold;'><A href='?src=[REF(src)];help=1'>\[?\]</A></div></HTML>", "window=[name]")
+		if(finalized)
```
There is no literacy check here.

WarlockD replied:
Sorry my dyslexia kicked in there

WarlockD replied:
It should of been checked in the can_interact(mob/user) when someone interacts, is it not working?

lordpidey commented on tgui/packages/tgui/components/TextArea.js:
```diff
@@ -0,0 +1,188 @@
+
+
+const toInputValue = value => {
+  if (isFalsy(value)) {
```
I'm unfamiliar with JS conventions in SS13, but it seems a bit strange to import a library when you could just use `if(value) {`

WarlockD replied:
Its already included in inferno, it basicly checks if its undefined, null or false.

lordpidey replied:
>it basicly checks if its undefined, null or false.

So does `if(value) {`, and that isn't dependent upon us ensuring the external library does what it says it does.

lordpidey commented on tgui/packages/tgui/interfaces/PaperSheet.js:
```diff
@@ -0,0 +1,153 @@
+    pen_color = "black",
+  } = data;
+  const onInputHandler = (e, value) => {
+    if (value.length < 1000) {
```
We're going to need server side checks too.

The whole reason paper was removed is due to someone acting maliciously.  Front end checks aren't going to stop that.

WarlockD replied:
I was afraid of that.  Checking for javascript or bad html won't be a problem,  The issue I have been having is if you send like a 100k text file, it kills byond.  Is there a way to stop it from that side?

WarlockD replied:
That is, if someone used a javascript injection attack, it would do it.  But then you could do the same for any ui_act updates.

WarlockD replied:
Take me an hour or so to find the functions but I will get some back end stuff to kill incoming html

optimumtact replied:
it should just check length, what the content is shouldn't matter, since it's only going to be rendered on the clientside again

WarlockD replied:
Thats simple enough to do.  I will fix post haste.

fludd12 commented on code/game/machinery/telecomms/computers/message.dm:
```diff
@@ -470,8 +470,12 @@
-	info_links = info
+	info = {"
+              Daily Key Reset
+The new message monitor key is 'server.decryptkey'.
```
You'll want to keep the squarebrackets around `server.decryptkey`, since this is a variable insert.

WarlockD replied:
ah thanks  missed that one.  After doing about 20 of those files at 2 am my mind gets warped

WarlockD committed:
Fixed some misspellings nad bracket issues

WarlockD committed:
Put a limit on the incomming text too

TheMidnightRose commented:
Soon?

WarlockD commented:
Properly till the github fix by the looks of it.

TheMidnightRose commented:
And Git seems to be fixed now!

kriskog commented:
merge conflict

WarlockD committed:
Merge branch 'master' into master

WarlockD commented:
ugh, let me repull and rebuild stupid tgui

WarlockD committed:
Ok lets try tavas again

WarlockD commented:
Alrighty fixed

WarlockD commented:
Why tgi bundle...whhhy

WarlockD committed:
Merge branch 'master' into master

WarlockD commented:
Well that taught me, never do a pr on the main fork, it screws up your branches good with the bundle tgui.

LetterN commented:
Possible issue. You might be able to show images using this. It looks like you're using marked.js as the interpreter for the markdown.

<https://marked.js.org/demo>

Here is a test case, it works on the marked.js demo.
```markdown
## Some text

[![fixthis](https://avatars1.githubusercontent.com/u/1363778?s=32&v=4)](https://www.youtube.com/watch?v=dQw4w9WgXcQ)
```
Outcome:

## Some text

[![fixthis](https://avatars1.githubusercontent.com/u/1363778?s=32&v=4)](https://www.youtube.com/watch?v=dQw4w9WgXcQ)

LetterN commented on tgui/packages/tgui/interfaces/PaperSheet.js:
```diff
@@ -0,0 +1,153 @@
+  // return sanitizer(marked(sanitizer(value),
+  //   { breaks: true, smartypants: true });
+  return sanitizer(marked(value,
+    { breaks: true, smartypants: true }));
```
```suggestion
    { breaks: true, smartypants: true, sanitize: true }));
```
I recommend turning it's own sanitizer on.

Arturlang commented:
Yeah, this kind of stuff is possible atm 
[![fixthis](https://avatars1.githubusercontent.com/u/1363778?s=200&v=4)](https://www.youtube.com/watch?v=dQw4w9WgXcQ)

![image](https://user-images.githubusercontent.com/24881678/83112946-9b283780-a0cf-11ea-8a02-38412ff1b0a2.png)

WarlockD commented:
At Least that's easy enough to fix, I just need to disable it in marked js.

WarlockD committed:
Fixed linking outside of byond

Also upgraded to marked 1.1.0 and sanatized the text when it is sent to byond.

WarlockD committed:
Merge branch 'master' into master

WarlockD commented:
Please try more stuff to break this.  I fixed the links to just display the raw marked [!] link.  Once the asset update is up, might change the links or url's to rewrite the asset sprite sheet.

WarlockD committed:
Update yarn.lock

WarlockD committed:
Merge branch 'master' of https://github.com/WarlockD/tgstation

WarlockD committed:
Update tgui.bundle.js

Cyberboss commented:
Testable state?

Arturlang commented:
I dont think stamps shows up inside the UI

WarlockD commented:
> Testable state?

Yea.  Believe it or not its running in the linux test server.  The one currenting going 100+ hours hah.

WarlockD committed:
Fixed it so ALL html is purged

WarlockD committed:
Of course stuipd bundle

WarlockD committed:
Ok lets try clearn it

WarlockD committed:
Ok, take two lets try this again

WarlockD commented:
I hate myself more than I hate github.  Can someone please merge tgui.bundle.js from the branch to this pull?  I have no idea how I could make a branch in this upstream.

WarlockD commented:
Killing my fork and going to rebuild it with a proper branch
